File: packages/connector/src/services/toggl-track/sync-time-entries.ts
/**
 * Toggl Track Time Entries Sync (Track API v9)
 *
 * Daily sync - fetches time entries for specified days and saves to raw layer.
 * Includes running entries (duration < 0).
 */

import { upsertRaw, RawRecord } from "../../db/raw-client.js";
import { setupLogger } from "../../lib/logger.js";
import { fetchTimeEntries } from "./api-client.js";

const logger = setupLogger("toggl-time-entries");

const TABLE_NAME = "toggl_track__time_entries";
const API_VERSION = "v9";

// Types
export interface SyncResult {
  success: boolean;
  count: number;
  elapsedSeconds: number;
}

/**
 * Convert API response to RawRecord
 */
function toRawRecord(entry: Record<string, unknown>): RawRecord {
  return {
    sourceId: String(entry.id),
    data: entry,
  };
}

/**
 * Format date as YYYY-MM-DD
 */
function formatDate(date: Date): string {
  return date.toISOString().slice(0, 10);
}

export interface SyncOptions {
  /** Number of days to sync (from today backwards). Ignored if start is provided. */
  days?: number;
  /** Start date (YYYY-MM-DD). If provided, days is ignored. */
  start?: string;
  /** End date (YYYY-MM-DD). Defaults to tomorrow. */
  end?: string;
}

/**
 * Sync time entries
 *
 * @param options - Sync options (days, start, end)
 * @returns Sync result
 */
export async function syncTimeEntries(options: SyncOptions = {}): Promise<SyncResult> {
  const startTime = performance.now();

  // Calculate date range
  let startDate: Date;
  let endDate: Date;

  if (options.start) {
    // Use explicit start date
    startDate = new Date(options.start);
    if (options.end) {
      endDate = new Date(options.end);
    } else {
      // Default end = tomorrow
      endDate = new Date();
      endDate.setDate(endDate.getDate() + 1);
    }
    logger.info(`Starting Toggl time entries sync (${options.start} to ${formatDate(endDate)})`);
  } else {
    // Use days (backwards from today)
    const days = options.days ?? 3;
    endDate = new Date();
    endDate.setDate(endDate.getDate() + 1); // Include today
    startDate = new Date();
    startDate.setDate(startDate.getDate() - (days - 1));
    logger.info(`Starting Toggl time entries sync (${days} days)`);
  }

  const startStr = formatDate(startDate);
  const endStr = formatDate(endDate);

  try {
    // Fetch from API
    logger.info(`Fetching time entries (${startStr} to ${endStr})...`);
    const entries = await fetchTimeEntries(startStr, endStr);
    logger.info(`Fetched ${entries.length} time entries`);

    // Save to raw layer
    let count = 0;
    if (entries.length > 0) {
      const records = entries.map(toRawRecord);
      const result = await upsertRaw(TABLE_NAME, records, API_VERSION);
      count = result.total;
    }

    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;
    logger.info(`Toggl time entries sync completed: ${count} entries in ${elapsed}s`);

    return {
      success: true,
      count,
      elapsedSeconds: elapsed,
    };
  } catch (error) {
    const elapsed = Math.round((performance.now() - startTime) / 10) / 100;
    logger.error(`Toggl time entries sync failed after ${elapsed}s: ${error}`);
    throw error;
  }
}

File: packages/connector/services/toggl_track/sync_time_entries.py
"""Toggl Track タイムエントリー同期（Track API v9）

日次同期用。指定日数分のタイムエントリーを取得してraw層に保存。
実行中エントリー（duration < 0）も取得可能。
"""

import time
from datetime import date, timedelta
from typing import Any, TypedDict

from pipelines.db.raw_client import upsert_raw, RawRecord
from pipelines.lib.logger import setup_logger
from pipelines.services.toggl_track.api_client import fetch_time_entries

logger = setup_logger(__name__)

TABLE_NAME = "toggl_track__time_entries"
API_VERSION = "v9"


class SyncResult(TypedDict):
    """同期結果"""
    success: bool
    count: int
    elapsed_seconds: float


def _to_raw_record(entry: dict[str, Any]) -> RawRecord:
    """APIレスポンスをRawRecordに変換

    source_idはエントリーIDを使用。
    """
    return RawRecord(
        source_id=str(entry["id"]),
        data=entry,
    )


async def sync_time_entries(days: int = 3) -> SyncResult:
    """タイムエントリーを同期（Track API v9）

    Args:
        days: 同期する日数（今日から遡る）

    Returns:
        同期結果
    """
    start_time = time.perf_counter()
    logger.info(f"Starting Toggl time entries sync ({days} days)")

    # 日付範囲を計算（end_dateは翌日にして当日分を確実に取得）
    end_date = date.today() + timedelta(days=1)
    start_date = date.today() - timedelta(days=days - 1)

    start_str = start_date.isoformat()
    end_str = end_date.isoformat()

    try:
        # API からデータ取得
        logger.info(f"Fetching time entries ({start_str} to {end_str})...")
        entries = await fetch_time_entries(start_str, end_str)
        logger.info(f"Fetched {len(entries)} time entries")

        # raw層に保存
        if entries:
            records = [_to_raw_record(e) for e in entries]
            result = await upsert_raw(TABLE_NAME, records, api_version=API_VERSION)
            count = result["total"]
        else:
            count = 0

        elapsed = round(time.perf_counter() - start_time, 2)
        logger.info(f"Toggl time entries sync completed: {count} entries in {elapsed}s")

        return SyncResult(
            success=True,
            count=count,
            elapsed_seconds=elapsed,
        )

    except Exception as e:
        elapsed = round(time.perf_counter() - start_time, 2)
        logger.error(f"Toggl time entries sync failed after {elapsed}s: {e}")
        raise
